### PR TITLE
Optional types

### LINKED ISSUE #89: Optional types
I want Some/None like rust

### PR DESCRIPTION
**Pull request by Brendonovich**

### CODE REVIEW

File: core/src/packages/utils.ts [hunk 1]
```diff
@@ -623,3 +545,106 @@ pkg.createNonEventSchema({
...
+    run({ ctx }) {
+      ctx.setOutput("equal", ctx.getInput("one") === ctx.getInput("two"));
+    },
+  });
+
+  pkg.createNonEventSchema({
+    name: `Unwrap Option (${key})`,
+    variant: "Pure",
+    run({ ctx }) {
+      ctx.getInput<Option<string>>("input").unwrap();
```
Comment (Hannah-GBS) [inline core/src/packages/utils.ts]: Not setting output here

File: core/src/packages/utils.ts [hunk 2]
```diff
@@ -623,3 +545,106 @@ pkg.createNonEventSchema({
...
+        id: "input",
+        type: types.option(type),
+      });
+      io.dataOutput({
+        id: "output",
+        type: types.bool(),
+      });
+    },
+    run({ ctx }) {
+      ctx.setOutput("out", ctx.getInput<Option<any>>("input").isSome());
```
Comment (Hannah-GBS) [inline core/src/packages/utils.ts]: output ID is "output" not "out"

File: core/src/packages/utils.ts [hunk 3]
```diff
@@ -623,3 +545,106 @@ pkg.createNonEventSchema({
...
+        id: "input",
+        type: types.option(type),
+      });
+      io.dataOutput({
+        id: "output",
+        type: types.bool(),
+      });
+    },
+    run({ ctx }) {
+      ctx.setOutput("out", ctx.getInput<Option<any>>("input").isNone());
```
Comment (Hannah-GBS) [inline core/src/packages/utils.ts]: output ID is "output" not "out"

File: core/src/packages/discord/gateway.ts [hunk 4]
```diff
@@ -137,7 +138,7 @@ pkg.createEventSchema({
     t.dataOutput({
       id: "username",
       name: "Username",
-      type: types.string(),
+      type: types.option(types.string()),
```
Comment (Hannah-GBS) [inline core/src/packages/discord/gateway.ts]: "nickname" should be the optional output, not "username"

File: core/src/packages/discord/gateway.ts [hunk 5]
```diff
@@ -161,14 +162,17 @@ pkg.createEventSchema({
...
-    console.log(data);
     ctx.setOutput("message", data.content);
     ctx.setOutput("channelId", data.channel_id);
     ctx.setOutput("username", data.author.username);
     ctx.setOutput("userId", data.author.id);
-    ctx.setOutput("nickname", data.member.nick);
-    ctx.setOutput("guildId", data.guild_id);
+    ctx.setOutput(
+      "nickname",
+      Option.new(data.member as any).map((v) => v.nick)
```
Comment (Hannah-GBS) [inline core/src/packages/discord/gateway.ts]: This currently resolves weirdly. Option is Some(String) passes True when there's no nickname. Changing it to `Option.new(data.member.nick as string | null))` fixes that but then there's a weird interaction with webhook messages, which trigger the Message Created event, but have no `member` or `roles` object.

Comment (Brendonovich) [reply core/src/packages/discord/gateway.ts]: Ahhhh i think i need to make a new Option from v.nick and use andThen instead of map, nice catch!

**Pull request merged**